### PR TITLE
WIP: Missile

### PR DESCRIPTION
**Pull request by TheZoq2**
Missile.

### CODE REVIEW

**Comment by TheZoq2** [comment]
No merge plz. Still has debug prints

**Review by Emiluren — APPROVED** [review]

File: libplen/src/gamestate.rs [hunk 1]
```diff
@@ -170,7 +161,7 @@ impl GameState {
         }
 
         self.projectiles.retain(
-            |projectile| projectile.is_done()
+            |projectile| !projectile.is_done()
```
Comment (TheZoq2) [inline libplen/src/gamestate.rs:164]: Has this been wrong the whole time?

File: libplen/src/player.rs [hunk 2]
```diff
@@ -165,6 +166,10 @@ impl Player {
         self.update_velocity_and_position(y_input, hurricane, delta_time);
         self.update_angular_velocity_and_rotation(x_input, delta_time);
         self.manage_powerups(delta_time);
+
+        send_line(
+            DebugLine::new(self.position, self.position + vec2(0., 100.)).rgb(255, 0, 0)
+        );
```
Comment (TheZoq2) [inline libplen/src/player.rs]: Remaining debug line

File: libplen/src/player.rs [hunk 3]
```diff
@@ -275,6 +280,23 @@ impl Player {
                 self.laser_charge_time = None;
             }
 
+            if self.weapon_is_wielded(PowerUpKind::Missile) && self.cooldown <= 0. {
+                let dir = self.rotation - std::f32::consts::PI / 2.;
+                self.cooldown = constants::PLAYER_COOLDOWN;
```
Comment (TheZoq2) [inline libplen/src/player.rs:280]: We may want to have a longer CD for missiles than gun. If so, this should probably be changed to something like if `self.last_shot > constants::PLAYER_COOLDOWN`.

We could do that in a separate PR though

File: libplen/src/player.rs [hunk 4]
```diff
@@ -287,13 +309,11 @@ impl Player {
                     self.id,
                     self.name.clone(),
                 );
-                (Some(ProjectileKind::from(new_bullet)), false)
-            } else {
-                (None, false)
+                return (Some(ProjectileKind::from(new_bullet)), false);
```
Comment (TheZoq2) [inline libplen/src/player.rs:307]: Consider renaming `new_bullet` to `new_projectile`

File: libplen/src/projectiles.rs [hunk 5]
```diff
@@ -169,42 +171,49 @@ impl Projectile for Missile {
...
                     direction_to.angle()
                 );
                 (direction_to, angle_to)
             })
-            .filter(|(_, angle_to)| {
-                 *angle_to > constants::MISSILE_LOCK_ANGLE
-            })
+            // .filter(|(_, angle_to)| {
+            //      *angle_to > constants::MISSILE_LOCK_ANGLE
+            // })
```
Comment (TheZoq2) [inline libplen/src/projectiles.rs]: Either implement this or remove it

File: libplen/src/projectiles.rs [hunk 6]
```diff
@@ -169,42 +171,49 @@ impl Projectile for Missile {
...
         };
 
+        send_line(
+            DebugLine::from_angle(self.position, self.angle, 50.).rgb(255, 255, 0)
+        );
+
+        send_line(
+            DebugLine::from_angle(self.position, self.angle + target_angle_diff, 50.).rgb(0, 255, 0)
+        );
+
```
Comment (TheZoq2) [inline libplen/src/projectiles.rs]: Lingoring debug prints

File: src/map.rs [hunk 7]
```diff
@@ -248,7 +251,7 @@ impl Map {
 
         Self::draw_killfeed(canvas, assets, game_state)?;
 
-        Self::draw_debug_lines(canvas, &game_state.debug_lines, camera_position)
+        Ok(())
```
Comment (TheZoq2) [inline src/map.rs:253]: Rather than removing debug drawing, we should remove the places that add lines to draw

File: src/server.rs [hunk 8]
```diff
@@ -112,6 +112,15 @@ impl Server {
...
 
+        send_line(
+            DebugLine::from_angle(vec2(0., 0.), 0., 100.).rgb(255, 0, 0)
+        );
+        send_line(
+            DebugLine::from_angle(vec2(0., 0.), 3.14 / 2., 100.).rgb(0, 255, 0)
+        );
+        send_line(
+            DebugLine::from_angle(vec2(100., 0.), 3.14 / 2., 100.).rgb(0, 255, 0)
+        );
```
Comment (TheZoq2) [inline src/server.rs]: Debug lines

**Review by TheZoq2 — COMMENTED** [review]
Me leaving a review on my own PR
image

File: src/map.rs (hunk 7 above)
Comment (Emiluren) [reply src/map.rs:253]: Oops my mistake

File: libplen/src/gamestate.rs (hunk 1 above)
Comment (Emiluren) [reply libplen/src/gamestate.rs:164]: You should know. You wrote it :P

File: libplen/src/player.rs (hunk 3 above)
Comment (Emiluren) [reply libplen/src/player.rs:280]: I think we should do it in a separate one

File: libplen/src/player.rs (hunk 2 above)
Comment (Emiluren) [reply libplen/src/player.rs]: Fixed

File: libplen/src/projectiles.rs (hunk 5 above)
Comment (Emiluren) [reply libplen/src/projectiles.rs]: I removed it

File: libplen/src/projectiles.rs (hunk 6 above)
Comment (Emiluren) [reply libplen/src/projectiles.rs]: They are lingOring no more :P

File: src/map.rs (hunk 7 above)
Comment (Emiluren) [reply src/map.rs:253]: Fixed

File: src/server.rs (hunk 8 above)
Comment (Emiluren) [reply src/server.rs]: Fixed

File: libplen/src/player.rs (hunk 4 above)
Comment (Emiluren) [reply libplen/src/player.rs:307]: Fixed

**Review by HannesTuhkala — APPROVED** [review]
Looks good
